src/interceptors/XMLHttpRequest/XMLHttpRequestOverride.ts

```diff
@@ -413,6 +415,25 @@ export const createXMLHttpRequestOverride = (
                 'original response headers (normalized)',
                 this._responseHeaders
               )
+            } else if (originalRequest.readyState === XMLHttpRequest.DONE) {
```

Comment:
What if we move this logic to `load`? I see the issue as this:

- Coupling updates to `this.status`, `this.statusText`, etc. with emitting the "response" event is a mistake. 

Instead, I think this also fixed the problem:

- Update internal response properties in `load`;
- Emit the response event only on `loadend` so it corresponds to successful responses. 

What do you think, @blaenk?

Reply:
Then we can also drop `finalized` internal state and rely on XHR respecting the spec when calling `load`.

Reply:
I think that was the other option outlined in #321. I defer to you on which one you prefer.

I went with this one for this pass of the PR because it seemed like it would handle it in every case, success or error, whereas the other approach I think I would have to handle it not only in `load` (success) but also in `error` and `abort` and maybe even `timeout`? I'm not sure. This way seemed more fundamental/universal. There's more info in #321 with links to the spec etc.

Curious what you think.

Reply:
As for only emitting the response in `loadend` I think that is captured by the current state of the PR but let me know if I misunderstood you.

Reply:
Afaik, `loadend` is called _always_, in both success and error scenarios, regardless of the error reason (rejection/timeout):

> The loadend event is fired when a request has completed, whether successfully (after [load](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/load_event)) or unsuccessfully (after [abort](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/abort_event) or [error](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/error_event)).
— [Source](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/loadend_event)

On the other hand, `load` is called only upon successful response. 

I believe what I'm proposing covers your issue rather well. There aren't any events at which response info is available prior to `load`. We populate that response info on `load`, and we finalize the response on `loadend`, which indicates a successful response. That should work.

Reply:
It seems there may have been a misunderstanding. I do understand that `loadend` runs in both cases. What I am referring to is your suggestion that we set the response info e.g. `.status` in `load`, my concern being that I believe that only runs on 'success', and so then unsuspecting code may have an `error` handler that checks `.status` or another one of those fields and in that case they will be uninitialized too.

Or is it the case that the `error` handler is for a kind of error which _wouldn't_ set those fields like `.status`, and they're only expected to be populated on success/`load`?

So to recap: I am concerned that setting the response info like `.status` only on `load` would mean that someone trying to read those fields from an `onerror` will see those fields as uninitialized. If so, then maybe response-field setting should also be set on other events as well?

Does that make sense now? If not let me know and I'll be happy to rephrase. If you don't believe that is an issue.

Reply:
By the way, to be clear, I am not against your suggestion, as it was one of the options I mentioned in #321, I'm simply bringing up that open question I had to see what you think, but I don't feel strongly one way or the other so I'm happy to implement it that way if you decide that's what you want!

Reply:
> we set the response info e.g. .status in load, my concern being that I believe that only runs on 'success',

That's the thing, `load` runs only on success but `loadend` runs in both success and failure. That's why the separation like I've suggested makes sense to me. I do believe that internally, what XHR does is exactly this:

```js
if (nextReadyState === this.DONE) {
  this.onloadend?(...)
}
```

> It may be the case that .status can only be expected to be set on a successful load

As we're speaking about this purely in the bypassed requests' case, any error/abort will not give us any status to infer, so when we do that in `loadend`, we will be doing nothing:

```js
originalRequest.addEventListener('loadend', () => {
  // If error happened, "originalRequest.status" is undefined/0/etc.
  // So we don't disrupt the original behavior. If anything, we respect it. 
  this.status = originalRequest.status
})
```

> I'm simply bringing up that open question I had to see what you think,

I think this proposal is great but it can be implemented in simpler terms (for us, as maintainers). As I see it, `readyState === DONE` check is the same as a `loadend` listener. 

The only concern here is the timing when these events run:

- `readystatechange` from X to DONE;
- and the dispatch of the `loadend` listeners.

I'm looking at some examples from the [XMLHttpRequest Specification](https://xhr.spec.whatwg.org/), and based on this error example I see that `readystatechange` always fries before any listeners:

> The request error steps for an [XMLHttpRequest](https://xhr.spec.whatwg.org/#xmlhttprequest) object xhr, event, and optionally exception are:
>
> Set xhr’s [state](https://xhr.spec.whatwg.org/#concept-xmlhttprequest-state) to done.
> 
> Unset xhr’s [send() flag](https://xhr.spec.whatwg.org/#send-flag).
> 
> Set xhr’s [response](https://xhr.spec.whatwg.org/#response) to a [network error](https://fetch.spec.whatwg.org/#concept-network-error).
> 
> If xhr’s [synchronous flag](https://xhr.spec.whatwg.org/#synchronous-flag) is set, then [throw](https://webidl.spec.whatwg.org/#dfn-throw) exception.
> 
> **[Fire an event](https://dom.spec.whatwg.org/#concept-event-fire) named [readystatechange](https://xhr.spec.whatwg.org/#event-xhr-readystatechange) at xhr.**
> 
> At this point it is clear that xhr’s [synchronous flag](https://xhr.spec.whatwg.org/#synchronous-flag) is unset.
> 
> If xhr’s [upload complete flag](https://xhr.spec.whatwg.org/#upload-complete-flag) is unset, then:
> 
> Set xhr’s [upload complete flag](https://xhr.spec.whatwg.org/#upload-complete-flag).
> 
> If xhr’s [upload listener flag](https://xhr.spec.whatwg.org/#upload-listener-flag) is set, then:
> 
> [Fire a progress event](https://xhr.spec.whatwg.org/#concept-event-fire-progress) named event at xhr’s [upload object](https://xhr.spec.whatwg.org/#upload-object) with 0 and 0.
> 
> [Fire a progress event](https://xhr.spec.whatwg.org/#concept-event-fire-progress) named [loadend](https://xhr.spec.whatwg.org/#event-xhr-loadend) at xhr’s [upload object](https://xhr.spec.whatwg.org/#upload-object) with 0 and 0.
> 
> [Fire a progress event](https://xhr.spec.whatwg.org/#concept-event-fire-progress) named event at xhr with 0 and 0.
> 
> **[Fire a progress event](https://xhr.spec.whatwg.org/#concept-event-fire-progress) named [loadend](https://xhr.spec.whatwg.org/#event-xhr-loadend) at xhr with 0 and 0**.

This means if we update internal properties on `loadend`, they won't yet be available in the preceding `readystatechange`:

```js
xhr.addEventListener('readystatechange', () => {
  if (xhr.readyStateChange === xhr.DONE) {
    console.log(xhr.status, xhr.statusText)
    // Will these be 0? 
  }
})
```

@blaenk, would you have a minute to see how regular XMLHttpRequest behaves in this regard? Are the updated response-related properties available in `readystatechange` on `DONE` immediately?

Reply:
I investigate some more and it seems like `onerror` for example is for network-level errors, so that definitely wouldn't set `status` since `status` assumes that there _was_ a request received. So in this case `load` being triggered on 'success' just means that a response was received, regardless of the http status code.

So it does make complete send to set those fields in `load` and I will make that change 👍🏼